File: 1주차/피로도.py
import sys
input = sys.stdin.readline

k = 80
dungeons = [[80,20], [50,40],[30,10]]

def dfs(k, dungeons, visit, ans=0):
    for i in range(len(dungeons)):
        mini, exh = dungeons[i]
        if k >= mini and not visit[i]:
            visit[i] = 1
            ans = max(dfs(k-exh, dungeons, visit, ans), sum(visit))
            visit[i] = 0
    return ans
    
def solution(k, dungeons):
    visit = [0] * len(dungeons)
    answer = dfs(k, dungeons, visit)
            
    return answer

print(solution(k, dungeons))

File: Oh/1주차/피로도.cpp
#include <bits/stdc++.h>

using namespace std;

int answer = 0;
bool visited[8] = {0};

void dfs(int cnt, int k, vector<vector<int>> &dungeons){
    if(cnt > answer) answer = cnt;
    
    for(int i=0; i<dungeons.size(); i++){
        if(!visited[i] && k - dungeons[i][0] >=0){
            visited[i] = 1;
          
            dfs(cnt+1, k - dungeons[i][1], dungeons);
            visited[i] = 0;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons){
    dfs(0, k, dungeons);
    return answer;
}

